### ISSUE TITLE
2022-03-03 Compute/GalleryRP StorageAccountType Swagger failing

### ISSUE THREAD

**Issue by JenGoldstrich** [Azure/azure-rest-api-specs#23588]
👋 Hello, 

In the 2022-03-03 Compute Gallery API the storageAccountType enum is defined with a different set of allowed values 

https://github.com/Azure/azure-rest-api-specs/blob/076362b3daf923937504ba93dcdf2959072eb0f9/specification/compute/resource-manager/Microsoft.Compute/GalleryRP/stable/2022-03-03/gallery.json#L2469 

https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/GalleryRP/stable/2022-03-03/gallery.json#L2560

This is causing an error for me when I am trying to add the 2022-03-03 compute APi to the SDK I am working on because of conflicting constants

```
error: -12T15:15:26.126Z [ERROR] Importer for Service "Compute".Importer for API Version "2022-03-03":      💥 Error: parsing Swagger files: parsing files in "../../swagger/specification/compute/resource-manager/Microsoft.Compute/GalleryRP/stable/2022-03-03": parsing definition: finding resources for tag "GalleryApplicationVersions": finding nested items yet to be parsed: appending model: appending constants: conflicting constant "StorageAccountType" with different values. First: map[PremiumLRS:Premium_LRS StandardLRS:Standard_LRS StandardZRS:Standard_ZRS]. Second: map[PremiumLRS:Premium_LRS StandardLRS:Standard_LRS StandardSSDLRS:StandardSSD_LRS StandardZRS:Standard_ZRS]
```

This seems to have been introduced in https://github.com/Azure/azure-rest-api-specs/pull/22255, looking at this PR I see the same change was also applied to the two previous API versions (2021-10-01 and 2022-01-03)

I see another PR here https://github.com/Azure/azure-rest-api-specs/pull/22829 which was opened to address this which seems like a logical solution but am unsure why it was closed

**Comment by Drewm3** [comment]
@sandeepraichura, could you please look into this issue.

**Comment by sandeepraichura** [comment]
Thanks for reporting it, we are currently looking into the issue and I'll keep you posted on the status in a couple of weeks.

**Comment by ArcturusZhang** [comment]
Hi @JenGoldstrich it is unfortunate that you get this issue because of the divergence of the enum values - we once had this issue when generating the "track 1" SDK which is an old version of the SDK generated from an old version generator and now it is deprecated.
The way we fix that is to use an additional patch layer of the tool to dynamically patch the generator to give one of the enums to have a different name: https://github.com/Azure/azure-rest-api-specs/blob/38311a16b5bfcb756164c6c48a64b0bc90d01a54/specification/compute/resource-manager/readme.csharp.md?plain=1#L39

In the new version of the SDKs, the generator could properly handle this case without issues by combining those two different enums. Side effect is that one enum has an additional value which is not valid right now. But this is coming from the thought that if those two enums have the same name, they are very probably to be the same thing, despite they diverged now, there is possibility that in the future, they no longer diverge.

We definitely do not want to just change it on the swagger because that would be a breaking change to all of our current SDKs, therefore I would like to ask if there could be any way to work around on your tool to properly handle cases like this?

**Comment by tombuildsstuff** [comment]
hey @ArcturusZhang 

Whilst we do have the ability to do _temporary_ workarounds within our importer - we intentionally only do these whilst the upstream data is being fixed (meaning they can be removed soon after) - so that's not something we can do as a long-term workaround without fixing the source data here unfortunately.

I'd actually argue the behaviour used by Track2 here is probably a bug, rather than a feature - since if multiple Enums are defined with the same name and differing values, how are users supposed to know which value is valid for which field? Instead, I'd suggest that these should be being caught by the OpenAPI Validator/Linting, since whilst I can understand it's possible for an Enum to be unintentionally used in the manner you're describing - as a consumer of that data, it's far more useful to have a missing Enum value (which can easily be added, since it's a non-breaking change) - verses being unable to determine which Enum values are valid. cc @mikekistler / @JeffreyRichter

Whilst obviously the semantics of that are larger than this specific issue, is it possible to get the source data fixed here and add an Autorest workaround for Track2 SDK's, so that the Swagger is correct here - and then look to determine if this behaviour should be updated?

Thanks!

**Comment by JeffreyRichter** [comment]
I tried reading this thread and I got lost. Can someone explain the whole thing to me in detail and then I can weigh in with an opinion as to how to fix/improve this.

**Comment by grizzlytheodore** [comment]
@JeffreyRichter  there are several places that define enum values for: storageAccountType 
but one of them has more enum values than the other, so SDK generation is failing for some. 

In this PR: https://github.com/Azure/azure-rest-api-specs/pull/23741 we are updating the one enum name to differentiate them

**Comment by grizzlytheodore** [comment]
are you trying to generate your own .NET SDK? 
if so you will need to resolve it by adding directives like this: 
https://github.com/Azure/azure-rest-api-specs/blob/38311a16b5bfcb756164c6c48a64b0bc90d01a54/specification/compute/resource-manager/readme.csharp.md?plain=1#L37

**Comment by tombuildsstuff** [comment]
@grizzlytheodore we're not generating a .net SDK - we're importing this data using our own tooling (which doesn't use AutoRest and [intentionally] doesn't support the AutoRest directives, since we favour fixing the data) - (see [this comment for more information](https://github.com/Azure/azure-rest-api-specs/issues/23588#issuecomment-1527681423)) - the resulting output is a Go SDK in the form of [`hashicorp/go-azure-sdk`](https://github.com/hashicorp/go-azure-sdk).

As mentioned in #23741, I've pulled this change down locally and confirm that #23741 fixes the issue we're seeing here - so thanks for that, what's needed to get that one merged?

---

@JeffreyRichter based on approving the breaking change in #23741 I have a feeling you've caught up on this one by this point, but if not:

We're currently parsing the Swagger files within each Service (`./specification/{service}`) by API version, meaning we group all files within a given API Version - unfortunately the updated folder structure being used is problematic for this approach, since Swagger files are no longer validated across the entire API Version, but on a per component level instead (e.g. Disks within Compute).

Whilst we can look into changing that to be able to import more granularly (at the component level, e.g. `compute/Disks/2022-03-03`) - that's something we'd need to look into and isn't something we support today.

As such the issue described above (and fixes in #23741) is that the Enum name for `StorageAccountType` isn't unique across the API Version - and support different values - meaning that our importer is erroring out since the same constant is defined with different values (which we treat as a data error, since these should be consistent if they're the same type - else should be unique types to indicate different options are available).

As mentioned above, I can confirm the fix for this in #23741 fixes this - so if it's possible for that to merged that would unblock us here.

Thanks,
Tom

**Comment by sandeepraichura** [comment]
Here's the PR that addresses this issue:  https://github.com/kjilla/azure-rest-api-specs/pull/1

**Comment by grizzlytheodore** [comment]
the fix has been merged. thank you everyone for your patience. closing this issue
